docs/Overview.md

```diff
@@ -1,96 +1,246 @@
 
+```go
+type CreateSource = fn (ExecutionContext) Source
+type CreateTransformation = fn (ExecutionContext) Transformation
+type CompilerType interface {
+    From() Language
+    To() Language
+}
 
-Both a query and a plan are represented by a DAG and describe an operation that needs to be performed.
-The difference is that a plan in addition to describing what the operation is, also describes how that operation will be performed.
-In short, a query describes what the operation is and a plan describes how that operation will be carried out.
+type Engine interface {
+    BindValue(id string, Value)
+    Prelude() Scope
+    
+    RegisterSource(id string, CreateSource)
+    RegisterTransformation(id string, CreateTransformation)
+    
+    Query(Representation) (Query, error)
+}
 
-## Relations
+type Scope interface {} // ...
+type Source interface {} // ...
+type Transformation interface {} // ...
+type Value interface {} // ...
+```
 
-Below is a high level description, using the Go language, of the relations between the different components and concepts of the query engine.
+A language is used to describe a query. It consists of words whose letters are taken from an alphabet.
+Letters can be of any type and complexity, from characters, to nodes in an Abstract Syntax Tree (AST). Various languages
+can be used to represent a query.
 
 ```go
-
-type Parser interface {
-    Parse(flux string) Query
+type Language interface {
+    Name() string
 }
+```
 
-// Query defines work to be performed on time series data and a result.
-// A query is represented as a directed acyclic graph (DAG).
+A query defines a unit of work for the engine. It comprises one or more pipelines.
+It contains its representation, its results, and execution statistics.
+
+```go
 type Query interface {
-    Nodes() []QueryNode
-    Edges() []Edge
+    Results() Results
+    Stats() map[string]interface{}
+    Representation() Representation
+}
+
+type Representation interface {
+    Visitor // a representation can be visited for compilation/execution.
+    Lang() Language // a representation makes sense in a certain language.
+    String() string // a visual representation of the query.
+}
+
+// The results for a Query.
+type Results interface {
+    ForEach(func (Result))
 }
 
-// QueryNode is a single step in the DAG of a query
-type QueryNode interface {
-    ID() NodeID
-    // More details about what the node does
+type Result interface {
+    Name() string
+    // ... methods to iterate other the contents of the result
 }
+```
 
-// NodeID uniquely identifies a node.
-type NodeID
+The executor executes a query expressed in some language and fills its results, provided an execution context.
+There can be multiple executors, one for each language supported by the engine.
+The engine is responsible for providing the executor with an
+execution context that embeds everything that is necessary for the execution of the query, included the registered values,
```

Comment:
With all due respect, I would like to disagree. I believe the engine is a concrete instance of an executor that takes as input a very specific representation that is the physical query plan.